### PR TITLE
Propagate correlationData to backend

### PR DESCRIPTION
**Pull request by mjudeikis**
Add correlationId to the backend

### CODE REVIEW

File: pkg/frontend/middleware/log.go [hunk 1]
```diff
@@ -52,9 +52,18 @@ func Log(baseLog *logrus.Entry) func(http.Handler) http.Handler {
 			r.Body = &logReadCloser{ReadCloser: r.Body}
 			w = &logResponseWriter{ResponseWriter: w, statusCode: http.StatusOK}
 
+			// correlationID is set by arm frontend. Local RP will not have this set
 			correlationID := r.Header.Get("X-Ms-Correlation-Request-Id")
```
Comment (mjudeikis) [inline pkg/frontend/middleware/log.go]: @jim-minter I assume `X-Ms-Correlation-Request-Id` one is set by ARM frontend, as in our current local dev its `""`. Should we set this if not set?

`X-Ms-Request-Id` is set by SDK, so we should generate only if API consumer is not standard and this field is empty?
WDYT?

Comment (jim-minter) [reply pkg/frontend/middleware/log.go]: x-ms-correlation-request-id: set by arm, don't set it, forward to backend for logging
x-ms-request-id: set by us, forward to backend for logging
x-ms-client-request-id: optionally set by client, forward to backend for logging

see https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md

File: pkg/frontend/middleware/log.go [hunk 2]
```diff
@@ -52,21 +52,27 @@ func Log(baseLog *logrus.Entry) func(http.Handler) http.Handler {
...
 
 			requestID := uuid.NewV4().String()
 			w.Header().Set("X-Ms-Request-Id", requestID)
 
 			fields := logrus.Fields{
-				"correlation_id": correlationID,
-				"request_id":     requestID,
+				"correlation_id":    correlationID,
+				"request_id":        requestID,
+				"client_request_id": r.Header.Get("X-Ms-Client-Request-Id"),
```
Comment (jim-minter) [inline pkg/frontend/middleware/log.go]: no need to log this

File: pkg/frontend/middleware/log.go (hunk 1 above)
Comment (mjudeikis) [reply pkg/frontend/middleware/log.go]: ok, so 3 fields to forward.  Maybe this?
```
doc.Metadata{
CorrelationRequestID
RequestID
ClientRequestID
}
```

File: pkg/frontend/middleware/log.go (hunk 2 above)
Comment (mjudeikis) [reply pkg/frontend/middleware/log.go]: `If the caller provides this header – the resource provider *must* log this with their traces to facilitate tracing a single request` Is this not applicable for us? 
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#client-request-headers

Comment (jim-minter) [reply pkg/frontend/middleware/log.go]: sure, we can do that one as well.  3 then.

**Comment by jim-minter** [comment]
lgtm; have pushed another commit to reduce repetition and move usage of log.WithFields into pkg/util/log - ptal?

**Comment by mjudeikis** [comment]
lgtm

**Pull request merged**